horizon/facts/update_subscriber.py

```diff
@@ -0,0 +1,95 @@
+import asyncio
+
+from loguru import logger
+from opal_client.data.updater import DataUpdater
+from opal_common.schemas.data import DataSourceEntry
+
+
+class DataUpdateSubscriber:
+    def __init__(self, updater: DataUpdater):
+        self._updater = updater
+        self._topic_events: dict[str, asyncio.Event] = {}
+
+    @property
+    def callbacks(self) -> dict[str, list[callable]]:
+        # Callbacks registered for PubSub client by topics
+        return self._updater._client._callbacks  # noqa
+
+    async def _on_message(self, topic: str = "", data=None) -> None:
+        if topic in self._topic_events:
+            logger.debug(f"Resolving subscriber event for topic: {topic}")
+            self._topic_events[topic].set()
+        else:
+            logger.info(f"No subscriber found for topic: {topic}")
+
+    def _get_event(self, topic: str) -> asyncio.Event:
+        if topic in self.callbacks:
+            if topic not in self._topic_events:
+                self._topic_events[topic] = asyncio.Event()
+
+            # Injecting the callback directly into the PubSub client, because subscribing to the topic is not possible
+            # after client is initialized. If the pubsub client does not already have callbacks for this topic, it is
+            # no longer possible to subscribe to it.
+            self.callbacks[topic].append(self._on_message)
+            return self._topic_events[topic]
+        else:
+            raise Exception(f"PubSubClient is not subscribed to topic: {topic}")
+
+    def _clear_event(self, topic: str) -> None:
+        if topic in self.callbacks:
+            self.callbacks[topic].remove(self._on_message)
+        if topic in self._topic_events:
+            self._topic_events.pop(topic)
+
+    async def wait_for_message(self, topic: str, timeout: float | None = None) -> bool:
+        logger.info(f"Waiting for message on topic: {topic}")
+        event = self._get_event(topic)
```

Comment:
the get event is outside of the try except and has "raise Exception" in some scenario.
this part is called before publishing and can result in not publishing the message to opal-server or am I wrong because it happens in a different task ?

Reply:
First, it is happening on another task, so we would just not wait for the message to arrive back. Second, the try-except is specifically for Timeout, it does not handle any other exception and let it propagate.